contracts/src/v0.8/keystone/KeystoneForwarder.sol

```diff
@@ -15,19 +15,55 @@ contract KeystoneForwarder is IForwarder, ConfirmedOwner, TypeAndVersionInterfac
   /// @param data the data that was received
   error InvalidData(bytes data);
 
+  error RepeatedSigner();
+
+  error WrongNumberOfSignatures();
+
+  error InvalidSigner();
+
   uint256 private constant SELECTOR_LENGTH = 4;
   uint256 private constant REPORT_LENGTH = 64;
 
   struct HotVars {
     bool reentrancyGuard; // guard against reentrancy
+    uint8 n;
+    uint8 f;
   }
 
   HotVars internal s_hotVars; // Mixture of config and state, commonly accessed
 
+  /// @notice Contains the signing address of each oracle
+  address[] internal s_signers;
+  mapping(address => uint8 index) internal s_signersMap; // TODO: is this more gas efficient than a for loop { eq }?
```

Comment:
I'd suggest starting with https://docs.openzeppelin.com/contracts/3.x/api/utils#EnumerableMap and doing gas optimizations as a separate step once we are feature complete.